### PR TITLE
[ray_client] Include multiple facets of the Ray API

### LINKED ISSUE #12426: Implement misc methods in client [10]
Here's a short list from https://docs.ray.io/en/master/package-ref.html

ray.is_initialized()
ray.kill()
ray.cancel()
ray.nodes()
ray.cluster_resources()
ray.available_resources()

### PR DESCRIPTION
**Pull request by barakmich**
Implements multiple metadata requests from the ray server, as well as termination commands (kill(), cancel())

Changes ClientObjectRef ids to be in line with ClientActorHandle ids (found bugs by cribbing termination tests)

Adds tests for these commands.

## Related issue number
Closes #12426

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by ericl** [comment]
Tried this out, a few comments:
- Can you add ray.is_initialized() from the issue?
- ray.cancel() / ray.kill() with incorrect arg should raise TypeError, instead gets odd error
- ray.cancel() / ray.kill() twice on the same arg raises KeyError on server, should be idempotent

Also:
- ClientObjectRef / ClientActorHandle is superlong: `ClientActorHandle(800595f7000000000000008c117261792e73657269616c697a6174696f6e948c196163746f725f68616e646c655f646573657269616c697a657294939443bd0a08de731761010000001210ffffffffffffffffffffffff010000001a3f0a146cf26e237cb0beb9435efb4508ac27484bbff3da120d3139322e3136382e352e313231188d9e02221401000000ffffffffffffffffffffffffffffffff220401000000323f123d0a085f5f6d61696e5f5f1201581a085f5f696e69745f5f222438343566323733352d323737372d343762382d383937312d3334383031626165643535383a14ffffffffffffffffde731761010000000100000042013194859452942e)`
Can we print the same thing as the server-side handle (real obj / actor id)?

File: python/ray/__init__.py [hunk 1]
```diff
@@ -109,7 +109,6 @@
     "get_runtime_context",
     "actor",
     "actors",
-    "workers",
```
Comment (ericl) [inline python/ray/__init__.py:112]: Why this change?

File: python/ray/experimental/client/api.py [hunk 2]
```diff
@@ -131,10 +166,136 @@ def call_remote(self, instance: "ClientStub", *args, **kwargs):
...
+            Information about the actors.
+        """
+        client_id = None
+        if actor_id is not None:
+            client_id = bytes.fromhex(actor_id)
+
+        return self.worker.get_cluster_info(
+            ray_client_pb2.ClusterInfoType.ACTORS, client_id=client_id)
+
+    def objects(self, object_ref: Optional[str] = None):
```
Comment (ericl) [inline python/ray/experimental/client/api.py:202]: Can we not expose this in the public API? Users always try to use this incorrectly; I believe it's been removed from the documentation.

File: python/ray/experimental/client/api.py [hunk 3]
```diff
@@ -131,10 +166,136 @@ def call_remote(self, instance: "ClientStub", *args, **kwargs):
...
     def __getattr__(self, key: str):
         if not key.startswith("_"):
             raise NotImplementedError(
                 "Not available in Ray client: `ray.{}`. This method is only "
                 "available within Ray remote functions and is not yet "
                 "implemented in the client API.".format(key))
         return self.__getattribute__(key)
+
+
+class ClientStateAPI:
```
Comment (ericl) [inline python/ray/experimental/client/api.py:260]: Let's not expose ray.state for now--- it's an undocumented API

**Review by ericl — CHANGES_REQUESTED** [review]
Couple requests to not expose undocumented APIs

**Comment by ericl** [comment]
@barakmich also, please make sure to use `Assign` and not `Request Review` on PRs.

File: python/ray/__init__.py (hunk 1 above)
Comment (barakmich) [reply python/ray/__init__.py:112]: Because it's broken, as I asked about the other day. There's no import in this file that satisfies `ray.workers()` -- so it's always been broken when called. So rather than expose something that never worked, I cleaned it up as I went by.

File: python/ray/experimental/client/api.py (hunk 2 above)
Comment (barakmich) [reply python/ray/experimental/client/api.py:202]: I didn't have the history on that; it is in the API so I figured it'd be nice to cover. Removal is easy enough.

File: python/ray/experimental/client/api.py (hunk 3 above)
Comment (barakmich) [reply python/ray/experimental/client/api.py:260]: That'd be fine, except `ray.nodes()` -- as was requested in the issue -- is part of the state API. Ditto also `cluster_resources()` and `available_resources()`. Let's sync on this.

**Comment by barakmich** [comment]
>   Can we print the same thing as the server-side handle (real obj / actor id)?

That would be difficult unless it specifically sends along a printable ID. Deserializing the ID on the client is piercing the veil to a ray import, which defeats the whole purpose.

There's a dirty way of doing this too -- just chopping out the bytes that are serialized because we can predict the offset -- but that's super fragile.

**Comment by ericl** [comment]
Why not send along the printable ID back to the client? It will be pretty
confusing if the handle shows up differently in the cluster than on the
client.

**Comment by barakmich** [comment]
I mean, it's already going to show up differently -- it's a ClientX -- but sure, if this is that important, why not.

File: python/ray/experimental/client/server/server.py [hunk 4]
```diff
@@ -23,19 +26,87 @@ def __init__(self, test_mode=False):
...
+        return json.dumps(data)
+
+    def Terminate(self, request, context=None):
+        if request.WhichOneof("terminate_type") == "task_object":
+            object_ref = cloudpickle.loads(request.task_object.id)
+            try:
+                obj = self.object_refs.get(object_ref.binary(), None)
+                if obj is None:
+                    return ray_client_pb2.TerminateResponse(ok=False)
+                ray.cancel(
```
Comment (ericl) [inline python/ray/experimental/client/server/server.py:70]: I'm not sure about this assumption that the object ref to cancel must be in self.object_refs. What if the object is created in a remote actor, and the ref is returned to the client somehow?

Could we remove the check here?

File: src/ray/protobuf/ray_client.proto [hunk 5]
```diff
@@ -42,20 +42,25 @@ message ClientTask {
   repeated Arg args = 4;
 }
 
+message RemoteRef {
+  bytes id = 1;
+  bytes handle = 2;
```
Comment (ericl) [inline src/ray/protobuf/ray_client.proto:47]: Can you document the semantics of these fields (can be later pr)

File: src/ray/protobuf/ray_client.proto [hunk 6]
```diff
@@ -65,15 +70,15 @@ message GetResponse {
 }
 
 message WaitRequest {
-  repeated bytes object_refs = 1;
+  repeated bytes object_handles = 1;
```
Comment (ericl) [inline src/ray/protobuf/ray_client.proto:73]: Handles is overloaded with actor handles, another term?

**Review by ericl — APPROVED** [review]

**Comment by ffbin** [comment]
@barakmich @ericl https://github.com/ray-project/ray/pull/12783 I have submit a pr to fix lint error, pls help take a look.
